src/sage/rings/species.py

```diff
@@ -0,0 +1,2390 @@
+        dompart = [[ZZ(C._domain_to_gap[e]) for e in b] for b in dompart]
+        orbits = sorted(G.Orbits().sage(), key=len, reverse=True)
+        for orbit in orbits:
+            if not any(set(orbit).issubset(b) for b in dompart):
+                o = [C._domain_from_gap[e] for e in orbit]
+                raise ValueError(f"All elements of orbit {o} must have the same sort")
+
+        S = SymmetricGroup(C.degree())
+
+        def new_dis():
+            """
+            Return a representative ``dis`` in the conjugacy
+            class of subgroups conjugate to ``C``, together with the
+            conjugating map ``mp``
+            """
+            # create a group conjugate to `C` whose orbits
+            # consist of contiguous sets of numbers, with larger
+            # orbits having smaller numbers, and compute a small
+            # generating set
+            mp = PermutationGroupElement([e for o in orbits for e in o]).inverse().gap()
+            H = PermutationGroup(G.SmallGeneratingSet().sage())
+            return H.conjugate(mp), mp
+
+        key = G.Size().sage(), tuple([len(o) for o in orbits])
+        S_gap = S.gap()
+        if key in parent._dis_cache:
+            lookup_dis = parent._dis_cache[key]
+            for dis in lookup_dis:
+                mp = libgap.RepresentativeAction(S_gap, G, dis)
+                if mp != GAP_FAIL:
+                    break
+            else:
+                dis, mp = new_dis()
+                lookup_dis.append(dis)
+        else:
+            dis, mp = new_dis()
+            lookup_dis = parent._dis_cache[key] = [dis]
+
+        dompart = [[ZZ(e ** mp) for e in b] for b in dompart]
+        mc = tuple([len(b) for b in dompart])
+        key = (mc, dis)
+        if key in parent._cache:
+            lookup = parent._cache[key]
+            if parent._arity == 1 and lookup:
+                assert len(lookup) == 1
+                return lookup[0]
+            domain = list(chain(*map(sorted, dompart)))
+            dis_gap = dis.gap()
+            dis_gens = dis_gap.GeneratorsOfGroup()
+            for elm in lookup:
+                # check whether the assignment to sorts given by
+                # dompart and by elm._dompart are the same
+                elm_domain = list(chain(*map(sorted, elm._dompart)))
+                mp = libgap.MappingPermListList(elm_domain, domain)
+                if all(g ** mp in dis_gap for g in dis_gens):
+                    return elm
+        else:
+            lookup = parent._cache[key] = []
+
+        dompart = tuple([frozenset(b) for b in dompart])
+        elm = WithPicklingByInitArgs.__classcall__(cls, parent, dis, dompart)
+        lookup.append(elm)
+        return elm
+
+    def __init__(self, parent, dis, domain_partition):
+        r"""
+        Initialize an atomic species.
+
+        INPUT:
+
+        - ``dis`` -- a directly indecomposable permutation group
+        - ``domain_partition`` -- a `k`-tuple of ``frozenset``s,
```

Comment:
```suggestion
        - ``domain_partition`` -- a `k`-tuple of ``frozenset`` entries,
```
This will not format properly.

Reply:
Actually, should all of this information should be moved to the (public facing) class level?